fig/cli/config.py

```diff
@@ -0,0 +1,97 @@
+from __future__ import unicode_literals
+from __future__ import absolute_import
+
+import six
+import errno
+import yaml
+import os
+import re
+
+from . import errors
+
+
+class Config:
+    """
+    Wraps our YAML configuration to introduce support for environment variables and defaults
+    """
+    def __init__(self, config_dict):
+        """
+        Takes a dict (usually the output of a yaml file) and stores it.
+        Importantly, it recurses any structured values to turn them into configs as well.
+        We run string interpolations on stringy values if they match the ${ENV_VAR} pattern.
+
+        :param config_dict: a dict with structured configuration details
+        :type config_dict: dict
+        """
+
+        self.config_dict = dict([(key, Config(value)) if type(value) == dict else (key, value)
+                                 for key, value in config_dict.items()])
+
+    def as_dict(self):
+        """
+        Recursively forces a preprocessed config back into dict type
+        This is much easier than rewriting all client code to use a different class,
+        while providing us with a optional upgrade path (implementing Config directly where possible)
+
+        :return: the config as represented in a dict format
+        :rtype: dict
+        """
```

Comment:
It seems like we're having to recurse over the entire config twice. Once in the constructor, and again in `as_dict()`. 

I understand the desire for a `Config` object, but I think to really gain much over using a dict, we'd need a well defined object like a `namedtuple`.

I like that all the changes in this PR are isolated, I definitely this we can wait on the well defined `Config` object.

So instead of this double recurse, maybe this could just be a function instead of a class? It would look a lot like as_dict() without the first clause checking for `Config` objects.